src/webob/acceptparse.py

```diff
@@ -255,11 +294,382 @@ class AcceptEncoding(Accept):
+                )
+            if isinstance(value, (tuple, list)):
+                result = []
+                for element in value:
+                    if isinstance(element, (tuple, list)):
+                        element = _item_qvalue_pair_to_header_element(
+                            pair=element
+                        )
+                    result.append(element)
+                header_str = ', '.join(result)
+            else:
+                header_str = str(value)
+        return header_str
+
+
+class AcceptLanguageValidHeader(AcceptLanguage):
+    """
+    Represent a valid ``Accept-Language`` header.
+
+    A valid header is one that conforms to :rfc:`RFC 7231, section 5.3.5
+    <7231#section-5.3.5>`.
+
+    We take the reference from the ``language-range`` syntax rule in :rfc:`RFC
+    7231, section 5.3.5 <7231#section-5.3.5>` to :rfc:`RFC 4647, section 2.1
+    <4647#section-2.1>` to mean that only basic language ranges (and not
+    extended language ranges) are expected in the ``Accept-Language`` header.
+
+    This object should not be modified. To add to the header, we can use the
+    addition operators (``+`` and ``+=``), which return a new object (see the
+    docstring for :meth:`AcceptLanguageValidHeader.__add__`).
+    """
+
+    # RFC 7231 Section 5.3.5 "Accept-Language":
+    # Accept-Language = 1#( language-range [ weight ] )
+    # language-range  =
+    #           <language-range, see [RFC4647], Section 2.1>
+    # RFC 4647 Section 2.1 "Basic Language Range":
+    # language-range   = (1*8ALPHA *("-" 1*8alphanum)) / "*"
+    # alphanum         = ALPHA / DIGIT
+    lang_range_re = r"""
+        \*|
+        (?:
+        [A-Za-z]{1,8}
+        (?:-[A-Za-z0-9]{1,8})*
+        )
+    """
+    lang_range_n_weight_re = _item_n_weight_re(item_re=lang_range_re)
+    lang_range_n_weight_compiled_re = re.compile(
+        lang_range_n_weight_re,
+        re.VERBOSE
+    )
+    accept_language_compiled_re = _list_1_or_more__compiled_re(
+        element_re=lang_range_n_weight_re,
+    )
+
+    def __init__(self, header_value):
+        """
+        Create an :class:`AcceptLanguageValidHeader` instance.
+
+        :param header_value: (``str``) header value.
+        :raises ValueError: if `header_value` is an invalid value for an
+                            ``Accept-Language`` header.
+        """
+        self._header_value = header_value
+        self._parsed = list(self.parse(header_value))
+        self._parsed_nonzero = [
+            (language_range, qvalue) for language_range, qvalue in self.parsed
+            if qvalue
+        ]
+
+    @property
+    def header_value(self):
+        """(``str``) The header value."""
+        return self._header_value
+
+    @property
+    def parsed(self):
+        """
+        (``list``) Parsed form of the header.
+
+        A list of (language range, quality value) tuples.
+        """
+        return self._parsed
+
+    @classmethod
+    def parse(cls, value):
```

Comment:
Can I ask what changes would you actually like to see? You may remember that I asked if we should use an abstract base class to enforce the interface, so that was what I was trying to do. The api of the three classes are already very similar. But I am truly not understanding what you are asking for. I need most to understand what @mmerickel  meant by

> 'these classes should be implementation details just like the old design of AcceptLanguage and NilAccept'

and 

> > isn't NilAccept very much an implementation detail

> Yes, it was an implementation detail - and I hope we can do something similar here.

Because it is working the same now as it did before, only with different classes: it was `AcceptLanguage` and `NilAccept` before, and `AcceptLanguageValidHeader`, `AcceptLanguageNoHeader`, and `AcceptLanguageInvalidHeader` now. `NilAccept` is not hidden within `AcceptLanguage`, and there is no requirement I can see that `AcceptLanguage` and `NilAccept` have the same api (for example, `NilAccept` does not have `.parse`. And `.quality` had different parameters, which we just removed.)

If we go into the Python interpreter and look at the object we get with `request.accept_language`, depending on the header value we supply, we would get either an `AcceptLanguage` or a `NilAccept`: `NilAccept` is not hidden. If I wanted to know what attributes and methods were available when I saw the `NilAccept`, I would look up the documentation for `NilAccept`, not `AcceptLanguage`.

If you'd like the three classes to have the exact same API, I can make that happen. But they would still identify as the individual subclasses, like they are in the PR now, and like they were before this PR.

Do you want this:

> It would return one of 3 different instances that conform to the same api.

Which is basically what it is now, but making sure the api is exactly the same across the three subclasses, and adding an attribute that identify the object as being in one of the three states instead of using `isinstance()`?
 
Or this:

> AcceptLanguageValidHeader would basically move wholesale into AcceptLanguage if I understand @mmerickel's comments correctly, and then the sub-classes that are NoHeader or InvalidHeader would have .parse() and others raise ValueError or some other exception?

Or, as I mentioned earlier, hide all three classes within an `AcceptLanguage` class by, say, proxying attribute and method access? (Which is what I think of as truly hiding the implementation detail: the object would identify as `AcceptLanguage` at all times.)

Please, help me understand the change you would like to see.